File: app/src/main/java/com/lmyby/ankihelper/data/dict/DictionaryRegister.java
package com.lmyby.ankihelper.data.dict;

import android.content.Context;

import com.lmyby.ankihelper.MyApplication;
import com.lmyby.ankihelper.data.ai.AIDictionaryConfig;
import com.lmyby.ankihelper.data.ai.AIConfigRepository;

import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by liao on 2017/4/27.
 */

public class DictionaryRegister {
    //在这里注册词典类
    private static Class<?>[] classList = new Class<?>[]{
            Oalde10.class,
            Cdepe4.class,
            Maldpe.class
    };

    private static List<IDictionary> dictList;

    public Class[] getDictionaryClassArray() {
        return classList;
    }

    public static List<IDictionary> getDictionaryObjectList() {
        dictList = new ArrayList<>();
        
        // Add static dictionary classes
        for (Class<?> c : classList) {
            try {
                dictList.add(
                        (IDictionary) c.getDeclaredConstructor(Context.class).newInstance(MyApplication.getContext())
                );
            } catch (NoSuchMethodException nsme) {
                // Handle missing constructor
            } catch (InstantiationException ie) {
                // Handle instantiation issues
            } catch (IllegalAccessException ie) {
                // Handle access issues
            } catch (InvocationTargetException ite) {
                // Handle invocation issues
            }
        }
        
        // Add AI Dictionary configurations from database
        List<AIDictionaryConfig> aiConfigs = AIConfigRepository.getAllAIDictionaryConfigs();
        for (AIDictionaryConfig config : aiConfigs) {
            dictList.add(new AIDictionary(config));
        }
        
        return dictList;
    }
}

File: app/src/main/java/com/lmyby/ankihelper/data/dict/DictionaryRegister.kt
package com.lmyby.ankihelper.data.dict

import android.content.Context
import com.lmyby.ankihelper.MyApplication
import com.lmyby.ankihelper.data.ai.AIConfigRepository

/**
 * Created by liao on 2017/4/27.
 * Converted to Kotlin as part of Phase 2 data model migration
 */
object DictionaryRegister {
    // Register dictionary classes here
    private val classList = arrayOf<Class<*>>(
        Oalde10::class.java,
        Cdepe4::class.java,
        Maldpe::class.java
    )

    fun getDictionaryClassArray(): Array<Class<*>> {
        return classList
    }

    @JvmStatic
    fun getDictionaryObjectList(): List<IDictionary> {
        val dictList = mutableListOf<IDictionary>()

        // Add static dictionary classes
        for (c in classList) {
            try {
                dictList.add(
                    c.getDeclaredConstructor(Context::class.java)
                        .newInstance(MyApplication.getContext()) as IDictionary
                )
            } catch (nsme: NoSuchMethodException) {
                // Handle missing constructor
            } catch (ie: InstantiationException) {
                // Handle instantiation issues
            } catch (ie: IllegalAccessException) {
                // Handle access issues
            } catch (ite: java.lang.reflect.InvocationTargetException) {
                // Handle invocation issues
            }
        }

        // Add AI Dictionary configurations from database
        val aiConfigs = AIConfigRepository.getAllAIDictionaryConfigs()
        for (config in aiConfigs) {
            dictList.add(AIDictionary(config))
        }

        return dictList
    }
}
